### PR TITLE
AuthN: New service to support multiple authentication providers for plugins

### PR DESCRIPTION
**Pull request by gamab**
**What is this feature?**

This PR is preparation work to later add a new `authProvider` and allow plugins to choose between OAuth2 based authentication and Service Account Token based authentication.

I'm trying to have small PRs to ease reviews.

This PR adds a new service that will call the correct `authProvider` (`OAuth2Server`, `ServiceAccounts`) upon registration.

Since the `ServiceAccounts` provider will only generate a token and no ecdsa key, I changed the structures to only have the common parts and added optional fields to represent the providers' specifics.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

https://github.com/grafana/grafana-enterprise/issues/5824

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/extsvcauth/models.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	Name string `json:"name"`
+	// Impersonation access configuration
+	// (this is not available on all auth providers)
+	Impersonation ImpersonationCfg `json:"impersonation"`
+	// Self access configuration
+	Self SelfCfg `json:"self"`
+	// Auth Provider that the client will use to connect to Grafana
+	AuthProvider AuthProvider `json:"authProvider"`
+	// Auth Provider specific config
+	AuthProviderCfg interface{} `json:"authProviderCfg"`
```
Comment (andresmgot) [inline pkg/services/extsvcauth/models.go:51]: I'd rather have strong typings here, it's easier to maintain. My suggestion is to have different, optional, fields, one for the current oauth provider and then add the new one for the service account.

File: pkg/services/extsvcauth/models.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+// ExternalServiceDTO represents the credentials that the ExternalService can use to connect to Grafana.
+type ExternalServiceDTO struct {
+	Name   string      `json:"name"`
+	ID     string      `json:"id"`
+	Secret string      `json:"secret"`
+	Extra  interface{} `json:"extra"` // Auth Provider specificities (ex: ecdsa key pair)
```
Comment (andresmgot) [inline pkg/services/extsvcauth/models.go:59]: same here, if the information in this `Extra` is relevant for the client I would have two separate fields (optional), one per auth type.

Also, I don't know if I am missing something but it doesn't seems that this struct is being marshalled or sent overthe wire so why the DTO naming and the JSON annotations?

**Review by andresmgot — COMMENTED** [review]
Looking good so far! I have a couple of comments: In general I'd avoid using `interface{}` for things that are expected to have a certain structure (not generic blobs) so we avoid type casting. Also, if structs are not exposed on an API (which I believe is the case here), there is no need to define JSON annotations and call them `DTO`s.

**Comment by gamab** [comment]
> I have a couple of comments: In general I'd avoid using `interface{}` for things that are expected to have a certain structure (not generic blobs) so we avoid type casting. 

I was expecting this comment :smile:. I don't have a strong opinion on this, I just didn't like specifics to pop up at the "generic" layer. I'll make the change :+1: 

> Also, if structs are not exposed on an API (which I believe is the case here), there is no need to define JSON annotations and call them DTOs.

I share the idea, why is it there then? It's because the initial idea was to have an endpoint to get all registered clients in the future.
That being said, as of now, we don't need it, I can remove it for now and we'll add it when needed.

File: pkg/services/extsvcauth/models.go (hunk 2 above)
Comment (gamab) [reply pkg/services/extsvcauth/models.go:59]: Back when we were investigating on External Service Authentication, we had an endpoint to register service in some of the branches. We removed it to merge it later. But as it's not needed now, let's just remove the DTO.

File: pkg/services/extsvcauth/oauthserver/external_service.go [hunk 3]
```diff
@@ -49,19 +33,27 @@ type ExternalService struct {
...
-	c2 := ExternalServiceDTO{
-		Name:        c.Name,
-		ID:          c.ClientID,
-		Secret:      c.Secret,
-		GrantTypes:  c.GrantTypes,
-		Audiences:   c.Audiences,
-		RedirectURI: c.RedirectURI,
+// ToExtSvc converts the ExternalService (used internally by the oauthserver) to extsvcauth.ExternalService (used outside the package)
+// If object must contain Key pairs, pass them as parameters, otherwise only the client PublicPem will be added.
+func (c *ExternalService) ToExtSvc(keys *extsvcauth.KeyResult) *extsvcauth.ExternalService {
```
Comment (gamab) [inline pkg/services/extsvcauth/oauthserver/external_service.go:38]: Now that's a bit confusing. I should probably rename the struct :sweat_smile:

**Review by andresmgot — APPROVED** [review]
LGTM now!

**Pull request merged**